openvpn-setup.py

```diff
@@ -0,0 +1,146 @@
+
+Requires root to run.
+'''
+
+from jinja2 import Template
+import os
+import subprocess
+import sys
+import shutil
+
+# Default settings for CA template
+default_ca_vars = {
+    'key_country': '"US"',
+    'key_province': '"MI"',
+    'key_city': '"Detroit"',
+    'key_org': '"LiquidInvestigations"',
+    'key_email': '"support@liquiddemo.org"',
+    'key_ou': '"node"',
+    'key_name': '"detroit-liquid.local"'
+}
+
+default_server_conf = {
+    'port': '1192',
+    'protocol': 'udp'
+}
+
+# set up some paths
+# templates
+template_dir = os.path.join(os.getcwd(),'templates')
+vars_template_file = os.path.join(template_dir, 'vars.j2')
+conf_template_file = os.path.join(template_dir, 'server.conf.j2')
+
+# certificate authority
+ca_output_dir = '/etc/openvpn/openvpn-ca'
+vars_output_file = os.path.join(ca_output_dir,'vars')
+
+# keys
+ca_keys_dir = '/etc/openvpn/openvpn-ca/keys'
+keyfiles = list(map(lambda x: os.path.join(ca_keys_dir, x),
+               ['ca.crt', 'ca.key', 'server.crt', 'server.key', 'ta.key', 'dh2048.pem']))
+
+# configuration
+conf_output_dir = '/etc/openvpn'
+conf_output_file = os.path.join(conf_output_dir, 'server.conf')
+
+def file_from_template(template_file, output_file, vars_dict):
+    '''
+    Creates a file from the specified Jinja2 template
+    '''
+    with open(template_file) as template_file, \
+            open(output_file, 'w') as output_file:
+
+        template = Template(template_file.read())
+        output_file.write(template.render(vars_dict))
+
+if __name__ == '__main__':
+    ''' note: this expects that ca_output_dir doesn't exist
+    '''
+    # Set up CA directory
+    print("setting up certificate authority directory")
+    subprocess.call(['make-cadir', ca_output_dir])
+
+    # Create 'vars' file from template
+    print("creating vars file from template")
+    file_from_template(vars_template_file, vars_output_file, default_ca_vars)
+
+    # 'source' the vars file.
+    # this creates a bash subprocess, sources the file, outputs the new
+    # environment, then writes it to os.environ, key by key.
+    # TODO: proper error handling here on Popen() and communicate()
+    print("setting OS environment from vars file")
+    os.chdir(ca_output_dir)
+    command = ['bash', '-c', 'source ' + vars_output_file + ' && env']
+
+    proc = subprocess.Popen(command, stdout = subprocess.PIPE)
+
+    for line in proc.stdout:
+        # TODO: are we OK to just 'decode('ascii').rstrip() here?'
+        (key, _, value) = line.decode('ascii').rstrip().partition("=")
+        os.environ[key] = value
+
+    proc.communicate()
+
+    # clean the keys directory, generate certificates and keys
+    os.chdir(ca_output_dir)
+    key_dir = os.environ['KEY_DIR']
+    key_size = os.environ['KEY_SIZE']
+    dh_file = os.path.join(key_dir, 'dh' + key_size + '.pem')
+    hmac_file = os.path.join(key_dir, 'ta.key')
+
+    print('cleaning up before key generation (running ./clean-all)')
+    subprocess.call(['./clean-all'])
```

Comment:
You can define a function to wrap subprocess details (e.g. https://github.com/liquidinvestigations/docker-arm/blob/b9a9377dc8b5853064485c7406152520cee2bbe8/upload#L7-L10):

```python
def run(args, **kwargs):
    args = [str(a) for a in args]
    print('+', ' '.join(args))
    return subprocess.run(args, check=True, **kwargs)
```

And call it like so:

```python
pkitool_env = {}
# ... populate pki_env with the output of `['bash', '-c', 'source ' + 'vars' + ' && env']`
run(['./clean-all'], env=pkitool_env)
```

Reply:
And, instead of calling `os.chdir(...)`, pass a `cwd` argument to `run`, whcih gets passed to subprocess.

Reply:
when I turn all of the subprocess() calls into individual functions, do you want them to return the value of run?

Reply:
Only if it's used for something.

Reply:
Client config looking good. Doesn't work with UDP but seems ok for TCP:

https://gist.github.com/alsegno/b1f14972ebc1d8c042a2bfb26d4bb6bf

More testing is needed, and revocation needs to be done.